cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,79 @@ cuML is thread safe so its functions can be called from multiple host threads if
 
 The implementation of cuML is single threaded.
 
+## Public cuML interface
+### Motivation
+Such a well-defined interface is useful for two main reasons:
```

Comment:
What about: `The public cuML interface is stateless because:` instead of `Such a well-defined interface is useful for two main resons:`

Reply:
Done.